pyomo/core/base/units_container.py

```diff
@@ -1285,33 +1343,222 @@ def check_units_equivalent(self, expr1, expr2):
+        ----------
+        src : float or Pyomo expression
+           The source value that will be converted
+        from_units : None or Pyomo units expression
+           The units on the src. If None, then this method will try
+           to retrieve the units from the src as a Pyomo expression
+        to_units : Pyomo units expression
+           The desired target units for the new value
+
+        Returns
+        -------
+           float : The converted value
+
+        """
+        src_pyomo_unit, src_pint_unit = self._get_units_tuple(src)
+        if from_units is None:
+            from_pyomo_unit, from_pint_unit = src_pyomo_unit, src_pint_unit
+        else:
+            from_pyomo_unit, from_pint_unit = self._get_units_tuple(from_units)
+            if src_pint_unit is not None and \
+               not _UnitExtractionVisitor(self)._pint_units_equivalent(src_pint_unit, from_pint_unit):
+                raise UnitsError('convert_value called with a src argument that contains units'
+                                 ' that do not agree with the from_units argument.')
+        to_pyomo_unit, to_pint_unit = self._get_units_tuple(to_units)
+
+        # convert the values
+        src_quantity = value(src) * from_pint_unit
+        dest_quantity = src_quantity.to(to_pint_unit)
+        return dest_quantity.magnitude
+
+    def drop_units(self, expr, expected_units):
+        """
+        This method returns a dimensionless expression by dividing
+        expr by its units.
+
+        This is to support, for example, calls to intrinsic functions
+        that may contain expressions with units. To ensure that the
+        units are correct before dropping them, you must provide the
+        expected units. If the expected units do not match the units
+        on expr, a UnitsError exception will be raised.
+
+        Parameters
+        ----------
+        expr : Pyomo expression
+           The expression that will be converted
+        expected_units : Pyomo units expression
+           The expected units on expr. If this does not match the 
+           units on expr, then a UnitsError exception will be raised.
+
+        Returns
+        -------
+           Pyomo expression
+
+        """
+        pyomo_expr_unit, pint_expr_unit = self._get_units_tuple(expr)
+        pyomo_expected_unit, pint_expected_unit = self._get_units_tuple(expected_units)
+        if not _UnitExtractionVisitor(self)._pint_units_equivalent(pint_expr_unit,
+                                                                   pint_expected_unit):
+            raise UnitsError("drop_units called, however the units on the expression: {}"
+                             " do not match the expected units: {}."
+                             "".format(str(pyomo_expr_unit), str(pyomo_expected_unit)))
+
+        return expr/pyomo_expr_unit
+
+    def check_constraint_data_units(self, condata):
+        if condata.equality:
+            if condata.lower == 0.0:
```

Comment:
We should add a comment here documenting the edge case on why we allow an apparent units inconsistency for equality expressions with a 0 RHS.

Reply:
Agreed. Done.